File: android/app/src/main/kotlin/com/thecityandthebike/data/api/AuthInterceptor.kt
package com.thecityandthebike.data.api

import com.thecityandthebike.data.local.TokenManager
import okhttp3.Interceptor
import okhttp3.Response
import javax.inject.Inject

class AuthInterceptor @Inject constructor(
    private val tokenManager: TokenManager
) : Interceptor {

    override fun intercept(chain: Interceptor.Chain): Response {
        val originalRequest = chain.request()

        // Skip auth header for login and register endpoints
        val path = originalRequest.url.encodedPath
        if (path.contains("/auth/login") || path.contains("/auth/register") || path.contains("/auth/refresh")) {
            return chain.proceed(originalRequest)
        }

        val token = tokenManager.getToken()
        if (token.isNullOrEmpty()) {
            return chain.proceed(originalRequest)
        }

        val authenticatedRequest = originalRequest.newBuilder()
            .header("Authorization", "Bearer $token")
            .build()

        return chain.proceed(authenticatedRequest)
    }
}

File: ios/Packages/TCATBNetworking/Sources/TCATBNetworking/AuthInterceptor.swift
import Foundation

struct AuthInterceptor: Sendable {
    private let tokenProvider: TokenProvider

    private static let unauthenticatedPaths = [
        "/auth/login",
        "/auth/register",
        "/auth/refresh",
    ]

    init(tokenProvider: TokenProvider) {
        self.tokenProvider = tokenProvider
    }

    func intercept(_ request: URLRequest) async -> URLRequest {
        let path = request.url?.path ?? ""
        for unauthPath in Self.unauthenticatedPaths {
            if path.contains(unauthPath) {
                return request
            }
        }

        guard let token = await tokenProvider.getAccessToken(), !token.isEmpty else {
            return request
        }

        var authenticatedRequest = request
        authenticatedRequest.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        return authenticatedRequest
    }
}
